src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -826,9 +858,39 @@ private[internal] trait TypeMaps {
     }
   }
 
+  /** A map to implement the `substSym` method. */
+  class SubstSymMap(override val from: List[Symbol], override val to: List[Symbol]) extends AbstractSubstSymMap
+
+  class StackedMutableSubstSymMap extends AbstractSubstSymMap {
+    // What follows is an implementation of stack, like a resizing ArrayList.
+    private[this] var _from: ArrayBuffer[List[Symbol]] = new ArrayBuffer[List[Symbol]](16)
+    private[this] var _to: ArrayBuffer[List[Symbol]]  = new ArrayBuffer[List[Symbol]](16)
+    _from.addOne(Nil)
+    _to.addOne(Nil)
+    override def from: List[Symbol] = _from(_from.size-1)
+    override def to  : List[Symbol] =   _to(_from.size-1)
+
+    def push(from: List[Symbol], to: List[Symbol]): Unit = {
+      _from.addOne(from)
+      _to.addOne(to)
+      reload
+    }
+    def pop(): Unit = {
+      _from.remove(_from.size-1)
+      _to.remove(_to.size-1) 
+      reload
```

Comment:
None of this code is atomic — that's fine if there's a global lock (which is supposed to be there), but racing calls to `pop` could easily cause the errors in this stacktrace.

https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/11075/testReport/junit/test.files/run_/reflection_sorted_members_scala/

Reply:
Instances of this class are being created in a `ReusableInstance`. Would that be enough for preventing race conditions?